File: solution/data_structure/5430/main.cpp
//Authored by : suin8
//Co-authored by : -
//Link : http://boj.kr/c58e7826ebb34f06b1a2e42cfd4e5deb

/*
시간이 빡빡해서 실제로 뒤집기, 삭제하기를 구현하면 안됩니다.
뒤집기는 실제로 뒤집는 것이 아닌 앞에서 뒤로 읽는걸 뒤에서 앞으로 읽게 하였고
삭제하기는 실제 삭제가 아닌 front포인터를 하나 증가시키는 식으로 구현했습니다.
*/

#include <bits/stdc++.h>

using namespace std;

void inputString(string str);
void reverseList();
void outputList();
bool deleteList();

vector<string> numlist;
bool _reverse = false;
int front, back;

int main() {
    ios::sync_with_stdio(false); cin.tie(nullptr);

    int T;
    cin >> T;

    while(T--){
        string p, arr;
        int n;
        cin >> p >> n >> arr;

        inputString(arr);

        //실행할 함수를보고 적절하게 구현
        bool error = false;
        for(int i = 0;i < p.length();i++){
            if(p[i] == 'R')
                reverseList();
            else {
                if(deleteList() == false) {
                    cout << "error" << '\n';
                    error = true;
                    break;
                }
            }
        }
        if(error == false) outputList();
        numlist.clear();
        _reverse = false;
    }

    return 0;
} 

//비어있는 상황 []도 출력을 할 경우가 있기 때문에
//for문 안에 마지막 숫자는 컴마(,)를 출력하지 않도록 하였습니다.
void outputList(){
    cout << '[';
    if(_reverse == false){
        for(int i = front;i <= back;i++){
            if(i == back) cout << numlist[i];
            else cout << numlist[i] << ",";
        }
    }
    else{
        for(int i = back;i >= front;i--){
            if(i == front) cout << numlist[i];
            else cout << numlist[i] << ",";
        }
    }
    cout << "]\n";
}

//뒤집혀 있는 상태에서는 back이 앞쪽이 됩니다.
//따라서 back크기를 하나 줄여줍니다.
//추가로 front > back 일때 error상황을 체크합니다.
bool deleteList(){
    if(front > back) return false;
    else {
        if(_reverse == false) front++;
        else back--;

        return true;
    }
}

//_reverse가 true면 뒤집혀있다 라고 생각합니다.
void reverseList(){
    if(_reverse == true) _reverse = false;
    else _reverse = true;
}

//문자열을 뒤에서부터 읽으며 숫자만 numlist에 넣습니다.
void inputString(string arr){
    string num = "";
    while(!arr.empty()){
        if(arr.back() >= '0' && arr.back() <= '9'){
            num += arr.back();
            arr.pop_back();
        }
        else{
            if(num.empty() == true)
                arr.pop_back();
            else{
                reverse(num.begin(), num.end());
                numlist.push_back(num);
                num = "";
                arr.pop_back();
            }
        }
    }
    
    reverse(numlist.begin(), numlist.end());

    front = 0;
    back = numlist.size() - 1;
}

File: solution/data_structure/5430/Main.java
//Authored by : suin8
//Co-authored by : -
//Link : http://boj.kr/fc75d4683aa44fd5be618ed3934ac634

/*
시간이 빡빡해서 실제로 뒤집기, 삭제하기를 구현하면 안됩니다.
뒤집기는 실제로 뒤집는 것이 아닌 앞에서 뒤로 읽는걸 뒤에서 앞으로 읽게 하였고
삭제하기는 실제 삭제가 아닌 front포인터를 하나 증가시키는 식으로 구현했습니다.
출력도 빠른 출력을 해야 시간초과가 나지 않습니다. 
*/

import java.util.*;
import java.io.*;

public class Main {
    static ArrayList <String> numlist = new ArrayList<String>();
    static boolean reverse = false;
    static int front, end;
    static BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

    public static void main(String[] args) throws IOException {
        FastReader rd = new FastReader();

        int T = rd.nextInt();

        for(int i = 0;i < T;i++) {
            String p = rd.nextLine();
            int n = rd.nextInt();

            // inputString => tokenizer로 구분자 제거 후 list에 삽입
            InputString(rd.nextLine());

            boolean error = false;
            for(int j = 0;j < p.length();j++) {
                if(p.charAt(j) == 'R')  
                    ReverseList(); // ListReverse => 뒤집기

                else {
                    if(front > end) {
                        // 에러상황
                        // front가 end보다 큰 상황은 
                        // front == end == 1 인 상황(list에 값이 하나)에서
                        // 하나를 지운 후 상황이다. 즉 지울 것이 없는 상태
                        bw.write("error\n");
                        bw.flush();
                        error = true;
                        break;
                    }
                    else // 지울 수 있는 문자가 있을 때
                        DeleteList(); // DeleteList => 0번 인덱스 삭제
                }
            }
            if(error == false) OutputList();

            numlist.clear();
            reverse = false;
        }
        bw.close();
    }

    static void OutputList() throws IOException{
        bw.write("[");
        // reverse가 false면 정상상태 => 앞(front) 뒤(end)방향으로 출력
        if(reverse == false) {
            for(int i = front;i <= end;i++) {
                if(i == end) bw.write(numlist.get(i));
                else bw.write(numlist.get(i) + ","); 
            }
        }
        // reverse가 true면 뒤집힌 상태 => 뒤(end) 앞(front)방향으로 출력
        else {
            for(int i = end;i >= front;i--) 
                if(i == front) bw.write(numlist.get(i));
                else bw.write(numlist.get(i) + ",");
        }
        bw.write("]\n");
        bw.flush();
    }

    static void DeleteList() {
        // 실제로 지우는 것이 아닌 가리키는 인덱스를 증가 또는 감소시킴
        // 정상 상태이면 앞 인덱스를 증가(앞을 삭제하게 되므로)
        // 뒤집힌 상태이면 뒤 인덱스를 감소(뒤가 가장 앞이 되므로) 
        if(reverse == false) front++;
        else end--;
    }

    static void ReverseList() {
        // 실제로 뒤집는 것이 아닌 뒤집혀있는지 아닌지만 판단
        // 후에 출력할때 출력 순서를 바꾼다
        if(reverse == true) reverse = false;
        else reverse = true;
    }

    static void InputString(String numstr) {
        StringTokenizer st = new StringTokenizer(numstr, "[ | , | ]");

        while(st.hasMoreTokens()) 
            numlist.add(st.nextToken());

        front = 0;
        end = numlist.size() - 1;
    }

    static class FastReader {
        BufferedReader br;
        StringTokenizer st;

        public FastReader() {
            br = new BufferedReader(new InputStreamReader(System.in));
        }

        String next() {
            while(st == null || !st.hasMoreElements()) {
                try {
                    st = new StringTokenizer(br.readLine());
                }
                catch (IOException e) {
                    e.printStackTrace();
                }
            }
            return st.nextToken();
        }

        int nextInt() { return Integer.parseInt(next()); }
        String nextLine() {
            String str = "";
            try {
                str = br.readLine();
            }
            catch (IOException e) {
                e.printStackTrace();
            }
            return str;
        }
    }
}
